chore: add the hometask description (#1)
adjust/fe-hometask: README.md

**Description:** Challenge for Adjust Frontend Developer

**Owner:** [adjust](https://github.com/adjust)

**Contacts:** frontend-hiring@adjust.com

---

## Welcome to our little coding exercise! 👋

Here you will be given the opportunity to play with JavaScript and Rockets in the same project. For this, we recommend you to block 2-3 hours of your time to complete it.

**Please make sure you have read this page entirely, before starting the challenge**. If you have any doubts, please reach out to us by [opening an issue](#how-to-request-help) as described at the bottom of this page.

**Importantly**, we would like to give all candidates taking this test the same opportunity to solve the exercise in their own way. Because of this, **please do not fork or share this repo (nor your solution) with anybody 🙏🏻**

<img align="center" src="https://i.imgur.com/ekyJNd9.jpg" width="600">

## Instructions

1. Push your solution to a **private repo** in your **personal Github account**
2. When you are ready for us to take a look, please add [adjust-frontend-hiring][adjust-frontend-hiring] (GitHub user) as collaborator

## General notes

- The final solution must be a working React application

> - You can use any library you see fit

Reviewer: Nitpick: 

```suggestion
- You can use any library if you see a good fit for it
```

> - Bringing in best practices such as tests, types, linting will be appreciated
> 

Reviewer: Some candidates leave a lot of unnecessary comments. We could add the following:

```suggestion
- Bringing in best practices such as tests, types, linting will be appreciated
- It's not necessary to heavily comment the code, please only leave comments where you think they are necessary to understand the related code. 
```

Author: > Some candidates leave a lot of unnecessary comments. We could add the following:

Tbh I don't clearly understand why this is a problem for the home task

Reviewer: It's not a problem per se, although the more "expectations" we set, the easier it will be for the candidates to align with our expectations.
For an abstract home task, they don't know if they should add notes on every step or not. If we explicitly mention that they don't need to spend time on this, they won't have doubts when submitting a solution. Otherwise they might think "Ok I better comment every line of code to show my understanding of it". We don't need this and can save them a bit of time.

---

## Exercise

### TASK #1 - JavaScript

Implement `prepareData` higher order function, which takes an object of filter params `{year, customerName}`, and returns a function that processes a list of missions, by only showing the ones that were launched in `year` and carried a payload belonging to `customerName`.

**Observations:**

> - Missions should appear in inverse chronological order (sort), with the exception of those that carried more payloads should appear first.

Reviewer: Suggestion:

```suggestion
- Missions should appear in inverse chronological order (sort), with the exception of those that carried more payloads should appear first. Consider adding a unit test for this case.
```

- Payloads are carried in the second stage of a rocket and they can belong to multiple customers.
- It doesn't matter to which `customerName` 'program' each payload belongs to as long as `customerName` is the customer.

**Example:**

Considering we have the list of missions from this [data fixture][data-fixtures], and the following filter params:

```js
{
  year: 2018,
  customerName: "NASA"
}
```

> The expected result should be:

Reviewer: We could also setup a code sandbox with test examples and ask them to copy the tests and make sure the test cases we mentioned are working correctly. So that we could open their solutions and check tests are green?

> The expected result should be:

Author: > We could also setup a code sandbox with test examples and ask them to copy the tests and make sure the test cases we mentioned are working correctly. So that we could open their solutions and check tests are green?

As far as I know, we have the following problem: a high rate of passing the home task challenge and a low rate of passing the pair coding interviews.
The idea was to increase the home task complexity.
We can still have tests since the challenge is the same. It's possible to have well-described "contracts" for both tasks and run tests during the home task review.

WDYT? 
Could you elaborate on why you think candidates should be asked to write tests directly? 🙏

> The expected result should be:

Reviewer: Mostly for the convenience of us, reviewers. It was convenient to use [elonmusk](https://github.com/adjust/elonmusk) script to check if our base expectations are intact - tests are passing.

It would be nice to still have this convenience – running a test script before reviewing the rest of the submission, so that we know the task is "completed" and we can proceed and evaluate other criteria (code quality, etc).

> The expected result should be:

Reviewer 2: ideally [with the ACs and design specs given], I would write e2e tests that elonmusk would inject to the candidate repo and run them

> The expected result should be:

Author: I'm sorry for the late response, had some blockers in the team to resolve
This week I'm preparing a talk for townhall
Can it wait until the end of the week?

> The expected result should be:

Reviewer: Hi there, I have opened an adjacent PR, where I tried to summarise some of the points left from the discussion here. Please take a look when you have a chance 🙏 

https://github.com/adjust/fe-hometask/pull/2

```js
[
  {
    flight_number: 62,
    mission_name: "Iridium NEXT Mission 6",
    payloads_count: 2,
  },
  {
    flight_number: 72,
    mission_name: "CRS-16",
    payloads_count: 1,
  },
  {
    flight_number: 64,
    mission_name: "CRS-15",
    payloads_count: 1,
  },
  {
    flight_number: 60,
    mission_name: "TESS",
    payloads_count: 1,
  },
  {
    flight_number: 59,
    mission_name: "CRS-14",
    payloads_count: 1,
  },
];
```

---

### TASK #2 - React & Hooks

Implement `RocketsList` component with the following specifications:

1. It takes `filerParams` object (with shape described in [task #1][task-1]) as a prop
2. It obtains a list of 'missions' from a [custom hook][custom-hook], which gets **the whole list of missions** from [SpaceX API][spacex-api] and processes them with `prepareData` function (from [task #1][task-1]) and `filterParams` prop
3. For each 'mission' obtained from the custom hook, it renders a string using [template literals][template-literals] with the following format: "#`flight_number` `mission_name` (`payloads_count`)"

> 4. While 'missions' are being fetched from API, it renders `"Loading..."` to screen

Reviewer: Nitpick:
```suggestion
4. While 'missions' are being fetched from API, it renders `"Loading..."` on the screen
```

> 5. If no 'missions' are obtained from the custom hook, it renders `"No data"` to screen

Reviewer: ```suggestion
5. If no 'missions' are obtained from the custom hook, it renders `"No data"` on the screen
```

# React & Rockets

**Description:** Challenge for Adjust Frontend Developer

**Owner:** [adjust](https://github.com/adjust)

**Contacts:** frontend-hiring@adjust.com

---

## Welcome to our little coding exercise! 👋

Here you will be given the opportunity to play with JavaScript and Rockets in the same project. For this, we recommend you to block 2-3 hours of your time to complete it.

> **Please make sure you have read this page entirely, before starting the challenge**. If you have any doubts, please reach out to us by [opening an issue](https://github.com/adjust/react-and-rockets#how-to-request-help) as described at the bottom of this page.

Reviewer 2: has to be changed

**Importantly**, we would like to give all candidates taking this test the same opportunity to solve the exercise in their own way. Because of this, **please do not fork or share this repo (nor your solution) with anybody 🙏🏻**

<img align="center" src="https://i.imgur.com/ekyJNd9.jpg" width="600">

## Instructions

1. You **MUST [clone this repo](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/cloning-a-repository-from-github/cloning-a-repository)** to a location of your choosing where you can work on your solution
2. Write your solutions
3. Push your solution to a **private repo** in your **personal Github account**

> 4. When you are ready for us to take a look, please add [adjust-frontend-hiring][adjust-frontend-hiring] (GitHub user) as collaborator

Reviewer 2: only #3-4 make sense now from this list

## Important rules

Your challenge submission will be disqualified and automatically rejected if any of the following happens:

> - Your **repo was not properly cloned** as specified on step #1 of Instructions. [See how to clone a repository here](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository). **Note:** If you cloned the repo properly, previous commits from this (original) repo, must be showing on your git history.

Reviewer 2: rm?

4. While 'missions' are being fetched from API, it renders `"Loading..."` to screen
5. If no 'missions' are obtained from the custom hook, it renders `"No data"` to screen

**Observations:**

- The list of missions should be re-processed if `filterParams` prop changes, **BUT no new API calls should happen** to [SpaceX API][spacex-api]
- You are not allowed to use any of the filter parameters provided by the [SpaceX API docs][spacex-api-docs], just fetch all available data and process it with `prepareData` function

**Example:**

Considering we pass the following `filterParams` as a prop:

```js
{
  year: 2018,
  customerName: "NASA"
}
```

The expected render should be:

```txt
#62 Iridium NEXT Mission 6 (2)
#72 CRS-16 (1)
#64 CRS-15 (1)
#60 TESS (1)
#59 CRS-14 (1)
```

## How to request help

If you have any doubts, you can reach out to us by simply [creating a GitHub issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-an-issue#creating-an-issue-from-a-repository) in your private repo.

> Describe your question(s) and [mention](https://docs.github.com/en/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#mentioning-people-and-teams) [adjust-frontend-hiring][adjust-frontend-hiring] in your comments (don't forget to add user as collaborator). We will then receive a notification and get back to you as soon as possible.

Reviewer 2: should be changed

Author: I didn't get the idea
Could you elaborate?

Reviewer 2: that was a reminder to update if we'll drop this repo. Giving it a second read, I think it's meant to be their repo, so all good

## Helpful links

- [SpaceX API Docs][spacex-api-docs]
- [Inviting collaborators to a personal repository][github-collaborators]

[spacex-api]: https://api.spacexdata.com/v3/launches/past
[spacex-api-docs]: https://docs.spacexdata.com/?version=latest#fce450d6-e064-499a-b88d-34cc22991bcc
[github-collaborators]: https://help.github.com/en/articles/inviting-collaborators-to-a-personal-repository

> [task-1]: https://github.com/adjust/react-and-rockets#task-1---javascript

Reviewer 2: ```suggestion
[task-1]: https://github.com/adjust/fe-hometask#task-1---javascript
```
